### ISSUE TITLE
Hook/Plugin System

### ISSUE THREAD

**Issue by sts** [ytti/oxidized#60]
We want to be notified for configuration changes at our devices. I cannot use GIT hooks for that purpose, since they don't trigger on a bare repository as its initialised by Oxidized.

There is [https://github.com/git-commit-notifier/git-commit-notifier](git-commit-notifier) which will create good looking change e-mails, which I would like to use for that purpose. 
We could probably just initiate it after creating the rugged commit [oxidized/output/git.rb](https://github.com/ytti/oxidized/blob/master/lib/oxidized/output/git.rb#L70).

What are your thoughts?

**Comment by MrRJ45** [comment]
This would be very nice to have. The ability to execute an external binary/script would be nice, with the ability to pass it arguments, such as the device name, ip, etc.

**Comment by SaaldjorMike** [comment]
:+1: for a hook system. I could really use a pre-commit hook to do some cleanup in configuration before committing any changes. Also a post-commit hook to push to a remote repository would also be of interest to me.

**Comment by laetrid** [comment]
Hello! Any advance in configuration changes notification?
Now I use http://git-scm.com/docs/git-send-email for this purpose, but it isn`t really convenient.

**Comment by ytti** [comment]
@sts can you test this? 

https://github.com/ytti/oxidized/tree/feature-hook

**Comment by aakso** [comment]
@sts and @ytti 
here is the example configuration for using the Exec hook:
```yaml
hooks:
  do_smth:
    type: exec
    events: [post_store, node_fail]
    cmd: 'env >> /tmp/oxiexec.txt'
    timeout: 120
    async: true
```

**Comment by aakso** [comment]
You can also implement hooks in Ruby by subclassing `Oxidized::Hook` and implementing `run_hook(ctx)` method.

**Comment by etern4ever** [comment]
Since I haven't replaced my production rancid with Oxidized yet (still waiting on the custom ports), I installed the feature-hook branch on my test box.  It is working well with a shell script I created to run git-commit-notifier against the git repo on post_store.

**Comment by ytti** [comment]
Released in 0.8.0
